app/app/src/main/java/de/stustapay/stustapay/repository/InfallibleRepository.kt

```diff
@@ -0,0 +1,101 @@
+package de.stustapay.stustapay.repository
+
+import de.stustapay.api.models.NewTicketSale
+import de.stustapay.api.models.NewTopUp
+import de.stustapay.stustapay.model.InfallibleApiRequest
+import de.stustapay.stustapay.model.InfallibleApiRequestKind
+import de.stustapay.stustapay.net.Response
+import de.stustapay.stustapay.storage.InfallibleApiRequestLocalDataSource
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class InfallibleRepository @Inject constructor(
+    private val dataSource: InfallibleApiRequestLocalDataSource,
+    private val topUpRepository: TopUpRepository,
+    private val ticketRepository: TicketRepository
+) {
+    private val scope: CoroutineScope =
+        CoroutineScope(Dispatchers.Default + CoroutineName("infallible"))
+    private lateinit var runner: Job
+
+    fun launch() {
+        runner = scope.launch {
+            dataSource.requests.collect {
+                var requests = it
+
+                var retry = true
+                var attempts = 3
+                while (retry && attempts > 0) {
+                    retry = false
+
+                    for ((id, request) in it) {
+                        when (process(request)) {
+                            InfallibleResult.Ok -> {
+                                dataSource.remove(id)
+                                requests = requests.filter { it.key != id }
+                            }
+
+                            InfallibleResult.Retry -> retry = true
+                        }
+                    }
+
+                    attempts -= 1
+                    delay(1000)
+                }
```

Comment:
what happens after the 3 attempts?

Reply:
The limit of 3 attempts is there to prevent a deadlock in request processing.
Processing of requests is triggered by a change in the data stored on the filesystem. However, due to the way Flows work, new data will only reach the collect lambda after it has exited. If we were to retry infinitely here, we would therefore never be able to process new requests.